File: builder.js
/**
 * builder
 */

const { MyBuilderBase } = require("./src/builder_base.js");
const { MyCBuilder } = require("./src/c_builder.js");
const { MyLuaBuilder } = require("./src/lua_builder.js");
const { MyHtmlBuilder } = require("./src/html_builder.js");
const { MyCodeBuilder } = require("./src/code_builder.js");

const { js, files, terminal, tools } = require("./src/tools.js");

const builder = {};
const UI_LENGTH = 48;
const builders = ["c", "lua", "html", "code"];
const tasks = [];

const MY_BUILDER_TEMPLATE = `
  const builder = require("builder");
  const task = builder.%s({
      name: "%s",
      debug: false,
      release: false,
      input: "./source.%s",
      output: "./target"
  }).start();
`;

function splitByUpper(word) {
    const res = [];
    const regex = /([A-Z][a-z]*)/g;
    let match;
    while ((match = regex.exec(word)) !== null) {
        res.push(match[0]);
    }
    return res;
}

function String_paddCenter(text, length, char) {
    return text.padStart((text.length + length)/2).padEnd(length)
}

async function builderInit() {
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
    console.log("|" + String_paddCenter("My Builder", UI_LENGTH, " ") + "|");
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
    console.log("| build.lua not found, creating ...");
    console.log("| please enter task name:");
    const taskName = await terminal.read_line();
    console.log("| please select task type:");
    const taskType = await terminal.read_selection(builders);
    const myBuilderText = string.format(MY_BUILDER_TEMPLATE, taskType, taskName, taskType, taskType);
    files.write('./build.lua', myBuilderText);
    console.log("| created!");
}

function builderHelp(obj) {
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
    console.log("|" + `${obj.__name__} Help`.center(UI_LENGTH, " ") + "|");
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
    const confKeys = {};
    const objFuncs = {};
    const arr = [obj.__class__, obj.__class__.__super__];
    for (const cls of arr) {
        for (const key in cls) {
            const val = cls[key];
            if (typeof val === 'function' && key.startsWith("set")) {
                const words = splitByUpper(key);
                const name = words.join("_").toLowerCase();
                confKeys[name] = true;
            }
            if (typeof val === 'function' && !key.startsWith("_")) {
                objFuncs[key] = true;
            }
        }
    }
    console.log('| keys:');
    for (const key in confKeys) {
        console.log("| * " + key);
    }
    console.log('| funs:');
    for (const key in objFuncs) {
        console.log("| * " + key);
    }
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
}

MyBuilderBase.createFunc = (obj, args) => {
    args = args || {};
    for (const [k, v] of Object.entries(args)) {
        js.assert(js.is_text(k), 'Invalid argument key for builder: ' + String(k));
        const wrds = k.toLowerCase().split("_");
        const name = wrds.reduce((accumulator, word) => {
            return accumulator + word.charAt(0).toUpperCase() + word.slice(1);
        }, "");
        const func = obj['set' + name];
        js.assert(js.is_function(func), 'Unknown argument key for builder: ' + String(k));
        func.call(obj, v);
    }
    obj.help = builderHelp;
    tasks.push(obj);
    return obj;
}

builder.C = MyCBuilder
builder.Lua = MyLuaBuilder
builder.Html = MyHtmlBuilder
builder.Code = MyCodeBuilder

builder.c = (...args) => new MyCBuilder(...args);
builder.lua = (...args) => new MyLuaBuilder(...args);
builder.html = (...args) => new MyHtmlBuilder(...args);
builder.code = (...args) => new MyCodeBuilder(...args);

builder.help = function () {
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
    console.log("|" + "builder help".center(UI_LENGTH, " ") + "|");
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
    console.log('| builders:');
    for (const v of builders) {
        console.log('|', "*", v);
    }
    console.log('| functions:');
    for (const k in builder) {
        if (!builders.includes(k)) {
            console.log('|', "*", k);
        }
    }
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
};

builder.tasks = function () {
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
    console.log("|" + "builder list".center(UI_LENGTH, " ") + "|");
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
    console.log('| tasks:');
    for (let i = 0; i < tasks.length; i++) {
        const obj = tasks[i];
        console.log('|', i + 1 + ".", obj.getName());
    }
    console.log("-" + "-".repeat(UI_LENGTH) + "-");
};

builder.find = function (name) {
    if (!js.is_text(name)) {
        throw new Error('Invalid task name for builder');
    }
    for (const obj of tasks) {
        if (obj.getName() === name) {
            return obj;
        }
    }
};

builder.tools = require("./src/tools.js");

if (require.main === module) {
    builderInit();
} else {
    module.exports = builder;
}

File: builder.py
"""
builder
"""

import os
import sys

currentPath = os.path.dirname(__file__)
sourcePath = os.path.join(currentPath, "src")
sys.path.append(sourcePath)

################################################################

import tools as _tools
from tools import *

from c_builder import MyCBuilder
from lua_builder import MyLuaBuilder
from html_builder import MyHtmlBuilder
from code_builder import MyCodeBuilder

################################################################

C = MyCBuilder
Lua = MyLuaBuilder
Html = MyHtmlBuilder
Code = MyCodeBuilder

class builder:
    c = MyCBuilder
    lua = MyLuaBuilder
    html = MyHtmlBuilder
    code = MyCodeBuilder
    tools = _tools
    pass

################################################################

UI_LENGTH = 48
builders = ["c", "lua", "html", "code"]
tasks = []

MY_BUILDER_TEMPLATE = """
import builder
task = builder.{0}({{
    'name': "{1}",
    'debug': False,
    'input':"./test.{2}",
    'output':"test",
}})
task.setLibs([])
task.setIcon('../../resources/test.ico')
task.start()
task.run()
"""

def split_by_upper(word):
    import re
    res = []
    regex = re.compile(r'([A-Z][a-z]*)')
    match = regex.findall(word)
    if match:
        res.extend(match)
    return res

def string_padd_center(text, length, char):
    return text.center(length, char)

async def builder_init():
    print("-" + "-" * UI_LENGTH + "-")
    print("|" + string_padd_center("My Builder", UI_LENGTH, " ") + "|")
    print("-" + "-" * UI_LENGTH + "-")
    print("| build.py not found, creating ...")
    print("| please enter task name:")
    task_name = await terminal.read_line()
    print("| please select task type:")
    task_type = await terminal.read_selection(builders)
    my_builder_text = MY_BUILDER_TEMPLATE.format(task_type.upper(), task_name, task_type)
    files.write('./build.py', my_builder_text, 'utf-8')
    print("| created!")

def builder_help(obj):
    print("-" + "-" * UI_LENGTH + "-")
    print("|" + f"{obj.__name__} Help".center(UI_LENGTH, " ") + "|")
    print("-" + "-" * UI_LENGTH + "-")
    conf_keys = {}
    obj_funcs = {}
    arr = [obj.__class__, obj.__class__.__base__]
    for cls in arr:
        for key in dir(cls):
            val = getattr(cls, key)
            if callable(val) and key.startswith("set"):
                words = split_by_upper(key)
                name = "_".join(words).lower()
                conf_keys[name] = True
            if callable(val) and not key.startswith("_"):
                obj_funcs[key] = True
    print('| keys:')
    for key in conf_keys:
        print("| * " + key)
    print('| funs:')
    for key in obj_funcs:
        print("| * " + key)
    print("-" + "-" * UI_LENGTH + "-")

def create_func(obj, args={}):
    for k, v in args.items():
        py.check(py.is_text(k), 'Invalid argument key for builder: ' + str(k))
        wrds = k.lower().split("_")
        name = "".join(word.capitalize() for word in wrds)
        func = getattr(obj, 'set' + name, None)
        py.check(callable(func), 'Unknown argument key for builder: ' + str(k))
        func(v)
    obj.help = builder_help
    tasks.append(obj)
    return obj

Globals.createFunc = create_func

def _builder_help():
    print("-" + "-" * UI_LENGTH + "-")
    print("|" + "builder help".center(UI_LENGTH, " ") + "|")
    print("-" + "-" * UI_LENGTH + "-")
    methods = [attr for attr in dir(builder) if callable(getattr(builder, attr)) and not attr.startswith("_")]
    print('| builders:')
    for k in methods:
        if k in builders:
            print('|', "*", k)
    print('| functions:')
    for k in methods:
        if k not in builders:
            print('|', "*", k)
    print("-" + "-" * UI_LENGTH + "-")

def _builder_tasks():
    print("-" + "-" * UI_LENGTH + "-")
    print("|" + "builder list".center(UI_LENGTH, " ") + "|")
    print("-" + "-" * UI_LENGTH + "-")
    print('| tasks:')
    for i, obj in enumerate(tasks):
        print('|', f"{i + 1}.", obj.getName())
    print("-" + "-" * UI_LENGTH + "-")


def _builder_find(name):
    if not py.is_text(name):
        raise ValueError('Invalid task name for builder')
    for obj in tasks:
        if obj.getName() == name:
            return obj

################################################################

builder.help = _builder_help
builder.tasks = _builder_tasks
builder.find = _builder_find

################################################################

if __name__ == "__main__":
    import asyncio
    asyncio.run(builder_init())
else:
    pass

